fiducial_slam/src/map.cpp

```diff
@@ -349,53 +407,80 @@ bool Map::lookupTransform(const std::string &from, const std::string &to,
+        tf2::Vector3 c = T_camBase.transform.getOrigin();
+        ROS_INFO("camera->base   %lf %lf %lf",
+                 c.x(), c.y(), c.z());
+        T_camBase.variance = 1.0;
+    }
+    else {
+        ROS_ERROR("Cannot determine tf from camera to robot\n");
+    }
+
+    if (lookupTransform(baseFrame, obs[0].T_camFid.frame_id_, time, T_baseCam.transform)) {
+        tf2::Vector3 c = T_baseCam.transform.getOrigin();
+        ROS_INFO("base->camera   %lf %lf %lf",
+                 c.x(), c.y(), c.z());
+        T_baseCam.variance = 1.0;
+    }
+    else {
+        ROS_ERROR("Cannot determine tf from robot to camera\n");
+        return numEsts;
+    }
 
     for (size_t i=0; i<obs.size(); i++) {
         Observation &o = obs[i];
 
         if (o.fid == 0) {
             // virtual fiducial 0 is at the origin
-            T_fid0Cam = o.T_fidCam;
+            T_fid0Base.setData(o.T_fidCam * T_camBase);
 
-            tf2::Vector3 t = T_fid0Cam.transform.getOrigin();
+            tf2::Vector3 t = T_fid0Base.transform.getOrigin();
             double r, p, y;
-            T_fid0Cam.transform.getBasis().getRPY(r, p, y);
+            T_fid0Base.transform.getBasis().getRPY(r, p, y);
 
             ROS_INFO("Pose MUL %lf %lf %lf %lf %lf %lf %lf",
-              t.x(), t.y(), t.z(), r, p, y, T_fid0Cam.variance);
+              t.x(), t.y(), t.z(), r, p, y, T_fid0Base.variance);
 
-            if (T_fid0Cam.variance < multiErrorThreshold) {
+            if (T_fid0Base.variance < multiErrorThreshold) {
                 useMulti = true;
             }
         }
         else if (fiducials.find(o.fid) != fiducials.end()) {
             const Fiducial &fid = fiducials[o.fid];
 
             tf2::Stamped<TransformWithVariance> p = fid.pose * o.T_fidCam;
+
             p.frame_id_ = mapFrame;
             p.stamp_ = o.T_fidCam.stamp_;
 
-            o.position = p.transform.getOrigin();
+            p.setData(p * T_camBase);
+	    o.position = p.transform.getOrigin();
             double roll, pitch, yaw;
             p.transform.getBasis().getRPY(roll, pitch, yaw);
 
-            /*
-               we print out all 6DOF of the camera pose in the world
-               frame as estimated from a single fiducial here.  
-               This is also the transform from base_link to camera
-               that would cause base_link to be at the origin of the
-               map frame.  This can be used to determine the pose of
-               the camera on the robot if a fiducial is correctly setup
-               in the map file
-            */
+            // Create variance according to how well the robot is upright on the ground
+            // TODO: Create variance for each DOF
+            // TODO: Take into account position according to odom
+            double s1 = o.position.z();
+            double s2 = 2.0 * sin(roll);
+            double s3 = 2.0 * sin(pitch);
+            p.variance = s1*s1 + s2*s2 + s2*s3 + 1e-4;
```

Comment:
I think that last term should be s3*s3 right?

Reply:
Yes, thanks for catching that.  Also the 1e-4 is the systemic error, that should use a param.